crates/state/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+    /// [`GlobalState::commit`].
+    pub async fn checkpoint(&mut self) -> Result<Commit> {
+        let dirty_changes = std::mem::take(&mut self.dirty_changes);
+
+        let mut signature = [0; 32];
+
+        for change in dirty_changes {
+            let old_value: Option<Vec<u8>> = self.get_by_hash(&change.0, None).await?;
+            if let Some(old_value) = old_value {
+                xor_signature(&mut signature, &hash_key_value_pair(&change.0, &old_value));
+            }
+            xor_signature(&mut signature, &hash_key_value_pair(&change.0, &change.1));
+
+            let key_hash = &change.0[..];
+            let value = &change.1[..];
+            sqlx::query!(
+                r#"
+                    INSERT INTO "values" ("commit_id", "key_hash", "value")
+                    VALUES (?1, ?2, ?3)
+                "#,
+                self.next_commit_id,
+                key_hash,
+                value
+            )
+            .execute(&self.sqlite)
+            .await?;
+        }
+
+        self.update_commit_signature(&signature).await
+    }
+
+    async fn commit_signature(&self, commit_id: i64) -> Result<Commit> {
+        let bytes = sqlx::query!(
+            r#"
+                SELECT "signature"
+                FROM "commits"
+                WHERE "id" = ?1
+            "#,
+            commit_id
+        )
+        .fetch_one(&self.sqlite)
+        .await?
+        .signature;
+        Ok(bytes.try_into().unwrap())
+    }
+
+    async fn update_commit_signature(&mut self, partial_signature: &[u8; 32]) -> Result<Commit> {
+        let mut signature: [u8; 32] = self.commit_signature(self.next_commit_id).await?;
+        xor_signature(&mut signature, partial_signature);
+        let signature_bytes = &signature[..];
+
+        sqlx::query!(
+            r#"
+                UPDATE "commits"
+                SET "signature" = ?1
+                WHERE "id" = ?2
+            "#,
+            signature_bytes,
+            self.next_commit_id
+        )
+        .execute(&self.sqlite)
+        .await?;
+
+        Ok(signature)
+    }
+
+    /// Returns the [`Commit`] signature of the last ever checkpoint; i.e.
+    /// persisted changes without dirty changes.
+    pub async fn current(&self) -> Result<Commit> {
+        self.commit_signature(self.next_commit_id).await
+    }
+
+    /// Returns the current root hash in the form of a [`Commit`].
+    ///
+    /// It returns an error in case there's any dirty changes: you must
+    /// [`rollback`](GlobalState::rollback) first.
+    pub async fn rewind(&mut self, _commit: &Commit) -> Result<()> {
+        if self.dirty_changes.is_empty() {
+            Ok(())
+        } else {
+            Err(GlobalStateError::DirtyChanges)
+        }
+    }
+
+    /// Erases all dirty changes from memory. Persisted data is left untouched.
+    pub fn rollback(&mut self) {
+        self.dirty_changes.clear();
+    }
+}
+
+#[cfg(test)]
```

Comment:
It's important that we simulate:

* Rewinding back to some `commit` and replaying the same transactions as previously but in a different order.
* Rewinding back to some `commit` and playing new transactions.
* Playing two transactions successfully, then fail the 3rd one and after that succeed the 4th transaction (when the 4th and 3rd transactions share part of the keys).

Reply:
On this point, we might need to make some decisions regarding forcing linear ordering at the API level. We do support querying of historical data, but rewinding and then playing new transactions requires soft-deleting the newer commits or something along that lines. This is one limitation we have, which shouldn't be a problem for reorgs (as you said, clients only use linear behavior) but it is for the kind of test you're describing. I'll think about this and get back to you.